[pt-br] Add docs/tasks/configure-pod-container/configure-pod-configmap.md (#35144)
kubernetes/website: content/pt-br/docs/tasks/configure-pod-container/configure-pod-configmap.md

Você pode usar [`kubectl describe`](/docs/reference/generated/kubectl/kubectl-commands/#describe) ou
[`kubectl get`](/docs/reference/generated/kubectl/kubectl-commands/#get) para recuperar informações
sobre um mapa de configuração.

#### Crie um mapa de configuração a partir de diretórios

Você pode usar `kubectl create configmap` para criar um mapa de configuração a partir de vários arquivos no mesmo diretório. Quando você está criando um mapa de configuração baseado em um diretório, o kubectl identifica arquivos cujo nome-base é uma chave válida no diretório e empacota cada um desses arquivos no novo mapa de configuração. Quaisquer entradas de diretório, exceto que arquivos regulares são ignorados (ex. subdiretórios, links simbólicos, dispositivos, pipes, etc).

Por exemplo:

```shell
# Criando o diretório local

mkdir -p configure-pod-container/configmap/

# Baixe os arquivos de amostra no diretório `configure-pod-container/configmap/` 

wget https://kubernetes.io/examples/configmap/game.properties -O configure-pod-container/configmap/game.properties
wget https://kubernetes.io/examples/configmap/ui.properties -O configure-pod-container/configmap/ui.properties

> # Crie o mapa de configuração

Reviewer: ```suggestion
# Crie o ConfigMap
```

kubectl create configmap game-config --from-file=configure-pod-container/configmap/
```

> O comando acima empacota cada arquivo, neste caso, `game.properties` e `ui.properties` no diretório `configure-pod-container/configmap/` dentro do mapa de configuração de nome game-config. Você pode exibir detalhes do mapa de configuração usando o seguinte comando:
> 

Reviewer: ```suggestion
O comando acima empacota cada arquivo, neste caso, `game.properties` e `ui.properties` no diretório `configure-pod-container/configmap/` dentro do ConfigMap de nome game-config. Você pode exibir detalhes do ConfigMap usando o seguinte comando:

```

```
Name:         game-config
Namespace:    default
Labels:       <none>
Annotations:  <none>

Data
====
game.properties:
----
enemies=aliens
lives=3
enemies.cheat=true
enemies.cheat.level=noGoodRotten
secret.code.passphrase=UUDDLRLRBABAS
secret.code.allowed=true
secret.code.lives=30
ui.properties:
----
color.good=purple
color.bad=yellow
allow.textmode=true
how.nice.to.look=fairlyNice
```

Os arquivos `game.properties` e `ui.properties` no diretório `configure-pod-container/configmap/` estão representados na seção `data` do mapa de configuração.

```shell
kubectl get configmaps game-config -o yaml
```
A saída é semelhante a esta:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  creationTimestamp: 2016-02-18T18:52:05Z
  name: game-config
  namespace: default
  resourceVersion: "516"
  uid: b4952dc3-d670-11e5-8cd0-68f728db1985
data:
  game.properties: |
    enemies=aliens
    lives=3
    enemies.cheat=true
    enemies.cheat.level=noGoodRotten
    secret.code.passphrase=UUDDLRLRBABAS
    secret.code.allowed=true
    secret.code.lives=30
  ui.properties: |
    color.good=purple
    color.bad=yellow
    allow.textmode=true
    how.nice.to.look=fairlyNice
```

> #### Crie mapa de configuração a partir de arquivos
> 
> Você pode usar `kubectl create configmap` para criar um mapa de configuração a partir de um arquivo individual, ou a partir de múltiplos arquivos.
> 
> Por exemplo,

Reviewer: ```suggestion
#### Crie um ConfigMap a partir de arquivos {#create-configmaps-from-files}

Você pode usar `kubectl create configmap` para criar um ConfigMap a partir de um arquivo individual, ou a partir de múltiplos arquivos.

Por exemplo,
```

Here, the title also requires an anchor for links to work properly. Also preferable to not translate ConfigMap as it is an API object.

```shell
kubectl create configmap game-config-2 --from-file=configure-pod-container/configmap/game.properties
```

> Produziria o seguinte mapa de configurações:

Reviewer: ```suggestion
Produziria o seguinte ConfigMap:
```

```shell
kubectl describe configmaps game-config-2
```

Onde a saída é semelhante a esta:

```
Name:         game-config-2
Namespace:    default
Labels:       <none>
Annotations:  <none>

Data
====
game.properties:
----
enemies=aliens
lives=3
enemies.cheat=true
enemies.cheat.level=noGoodRotten
secret.code.passphrase=UUDDLRLRBABAS
secret.code.allowed=true
secret.code.lives=30
```

> Você pode passar o argumento `--from-file` múltiplas vezes para criar um mapa de configuração a partir de múltiplas fontes de dados.

Reviewer: ```suggestion
Você pode passar o argumento `--from-file` múltiplas vezes para criar um ConfigMap a partir de múltiplas fontes de dados.
```

```shell
kubectl create configmap game-config-2 --from-file=configure-pod-container/configmap/game.properties --from-file=configure-pod-container/configmap/ui.properties
```

> Você pode exibir detalhes do mapa de configuração `game-config-2` usando o comando a seguir:

Reviewer: ```suggestion
Você pode exibir detalhes do ConfigMap `game-config-2` usando o comando a seguir:
```

```shell
kubectl describe configmaps game-config-2
```

A saída é semelhante a esta:

```
Name:         game-config-2
Namespace:    default
Labels:       <none>
Annotations:  <none>

Data
====
game.properties:
----
enemies=aliens
lives=3
enemies.cheat=true
enemies.cheat.level=noGoodRotten
secret.code.passphrase=UUDDLRLRBABAS
secret.code.allowed=true
secret.code.lives=30
ui.properties:
----
color.good=purple
color.bad=yellow
allow.textmode=true
how.nice.to.look=fairlyNice
```

> Quando o `kubectl` cria um mapa de configuração a partir de entradas que não são ASCII ou UTF-8, a ferramenta os coloca no campo `binaryData` do mapa de configuração, e não no `data`. Fontes de dados de texto e binário podem ser combinadas em um mapa de configuração.
> Se você quiser ver a chave `binaryData` (e seus valores) em um mapa de configuração, você pode executar `kubectl get configmap -o jsonpath='{.binaryData}' <name>`.
> 
> Use a opção `--from-env-file` para criar um mapa de configuração a partir de um arquivo de ambiente, por exemplo:

Reviewer: ```suggestion
Quando o `kubectl` cria um ConfigMap a partir de entradas que não são ASCII ou UTF-8, a ferramenta os coloca no campo `binaryData` do ConfigMap, e não no campo `data`. Fontes de dados de texto e binário podem ser combinadas em um ConfigMap.
Se você quiser ver o campo `binaryData` (e seus valores) em um ConfigMap, você pode executar `kubectl get configmap -o jsonpath='{.binaryData}' <name>`.

Use a opção `--from-env-file` para criar um ConfigMap a partir de um arquivo de ambiente, por exemplo:
```

```shell
# Os arquivos de ambiente contêm uma lista de variáveis de ambiente.
# Essas regras de sintaxe se aplicam:
#   Cada linha em um arquivo de ambiente deve estar em formato VAR=VAL.
#   Linhas começando com # (ex. comentários) são ignoradas.
#   Linhas em branco são ignoradas.

> #   Não há manuseio especial de aspas (ex. eles farão parte dos valores do mapa de configuração).

Reviewer: ```suggestion
#   Não há manuseio especial de aspas (ex. eles farão parte dos valores do ConfigMap).
```

onde a saída é semelhante a esta:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  creationTimestamp: 2017-12-27T18:36:28Z
  name: game-config-env-file
  namespace: default
  resourceVersion: "809965"
  uid: d9d1ca5b-eb34-11e7-887b-42010a8002b8
data:
  allowed: '"true"'
  enemies: aliens
  lives: "3"
```

Começando com Kubernetes v1.23, o `kubectl` suporta o argumento `--from-env-file` ser
especificado várias vezes para criar um mapa de configuração para múltiplas fontes de dados.

```shell
kubectl create configmap config-multi-env-files \
        --from-env-file=configure-pod-container/configmap/game-env-file.properties \
        --from-env-file=configure-pod-container/configmap/ui-env-file.properties
```

Produziria o seguinte mapa de configuração:

```shell
kubectl get configmap config-multi-env-files -o yaml
```

Onde a saída é semelhante a esta:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  creationTimestamp: 2017-12-27T18:38:34Z
  name: config-multi-env-files
  namespace: default
  resourceVersion: "810136"
  uid: 252c4572-eb35-11e7-887b-42010a8002b8
data:
  allowed: '"true"'
  color: purple
  enemies: aliens
  how: fairlyNice
  lives: "3"
  textmode: "true"
```

> #### Defina a chave a ser usada ao criar um mapa de configuração a partir de um arquivo

Reviewer: ```suggestion
#### Defina a chave a ser usada ao criar um ConfigMap a partir de um arquivo
```

    secret.code.lives=30
```

#### Criando um mapa de configuração a partir de valores literais

Você pode usar `kubectl create configmap` com o argumento `--from-literal` para definir um valor literal a partir da linha de comando:

```shell
kubectl create configmap special-config --from-literal=special.how=very --from-literal=special.type=charm
```

Você pode passar vários pares de chave-valor. Cada par fornecido na linha de comando é representado como uma entrada separada na seção `data` do mapa de configuração.

```shell
kubectl get configmaps special-config -o yaml
```

A saída é semelhante a esta:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  creationTimestamp: 2016-02-18T19:14:38Z
  name: special-config
  namespace: default
  resourceVersion: "651"
  uid: dadce046-d673-11e5-8cd0-68f728db1985
data:
  special.how: very
  special.type: charm
```

### Crie um mapa de configuração de um gerador

O `kubectl` suporta `kustomization.yaml` desde a versão 1.14.
Você também pode criar um mapa de configuração a partir de geradores, e então aplicá-lo na criação de objetos no `Apiserver`. 
Os geradores devem ser especificado em um `kustomization.yaml` dentro de um diretório.

> #### Gere o mapa de configuração a partir de arquivos
> 
> Por exemplo, para gerar um mapa de configuração a partir de arquivos `configure-pod-container/configmap/game.properties`

Reviewer: ```suggestion
#### Gere o ConfigMap a partir de arquivos

Por exemplo, para gerar um ConfigMap a partir de arquivos `configure-pod-container/configmap/game.properties`
```

```shell

> # Criando um arquivo kustomization.yaml com o `ConfigMapGenerator`

Reviewer: ```suggestion
# Crie um arquivo kustomization.yaml com um ConfigMapGenerator
```

> # Criando um arquivo kustomization.yaml com o `ConfigMapGenerator`

Reviewer: ```suggestion
# Crie um arquivo kustomization.yaml com um ConfigMapGenerator
```

> # Criando um arquivo kustomization.yaml com o `ConfigMapGenerator`

Reviewer: ```suggestion
# Crie um arquivo kustomization.yaml com um ConfigMapGenerator
```

cat <<EOF >./kustomization.yaml
configMapGenerator:
- name: game-config-4
  files:
  - configure-pod-container/configmap/game.properties
EOF
```

Aplique o diretório de kustomization para criar o objeto mapa de configuração.

```shell
kubectl apply -k .
configmap/game-config-4-m9dm2f92bt created
```

> Você pode verificar se o mapa de configuração foi criado, assim:

Reviewer: ```suggestion
Você pode verificar que o ConfigMap foi criado, assim:
```

```shell
kubectl get configmap
NAME                       DATA   AGE

> game-config-4-m9dm2f92bt   1      37s

Reviewer: This block is originally broken into two.

The first one is a fenced code block of type `` ```shell ``, like so:
```shell
kubectl get configmap
```

The second one is a fenced code block without a type (`` ``` ``), like so:
```console
NAME                       DATA   AGE
game-config-4-m9dm2f92bt   1      37s
```

That's because these two blocks are rendered differently in the docs. The translated version should keep this formatting too.

kubectl describe configmaps/game-config-4-m9dm2f92bt
Name:         game-config-4-m9dm2f92bt
Namespace:    default
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"v1","data":{"game.properties":"enemies=aliens\nlives=3\nenemies.cheat=true\nenemies.cheat.level=noGoodRotten\nsecret.code.p...

Data
====
game.properties:
----
enemies=aliens
lives=3
enemies.cheat=true
enemies.cheat.level=noGoodRotten
secret.code.passphrase=UUDDLRLRBABAS
secret.code.allowed=true
secret.code.lives=30

> Events:  <none>

Reviewer: Same comment as before applies here. Command and output should be broken down into two separate fenced code blocks, first one of type ` ```shell `, second one of empty type (` ``` `). There's also a missing connecting sentence prior to these two blocks.

```

> Observe que o nome gerado para o mapa de configuração tem um sufixo anexado, que é o `hash` do conteúdo. Isso garante que um
> novo mapa de configuração é gerado cada vez que o seu conteúdo é modificado.
> 
> #### Defina a chave a ser usada ao gerar um mapa de configuração a partir de um arquivo
> 
> Você pode definir uma chave que não seja o nome do arquivo para usar no gerador do mapa de configuração.
> Por exemplo, para gerar um mapa de configuração a partir de arquivos `configure-pod-container/configmap/game.properties`
> com a chave `game-special-key`

Reviewer: ```suggestion
Observe que o nome gerado para o ConfigMap tem um sufixo anexado, que é o _hash_ do conteúdo. Isso garante que um
novo ConfigMap é gerado cada vez que o seu conteúdo é modificado.

#### Defina a chave a ser usada ao gerar um ConfigMap a partir de um arquivo

Você pode definir uma chave que não seja o nome do arquivo para usar no gerador do ConfigMap.
Por exemplo, para gerar um ConfigMap a partir de arquivos `configure-pod-container/configmap/game.properties`
com a chave `game-special-key`
```

```shell
# Criando um arquivo kustomization.yaml com o ConfigMapGenerator
cat <<EOF >./kustomization.yaml
configMapGenerator:
- name: game-config-5
  files:
  - game-special-key=configure-pod-container/configmap/game.properties
EOF
```

Aplicar o diretório kustomization para criar o objeto mapa de configuração.

```shell
kubectl apply -k .
configmap/game-config-5-m67dt67794 created

> ```

Reviewer: This block should also be broken down into two blocks, one for command and one for output.

#### Gere mapa de configuração a partir de literais

Para gerar um mapa de configuração a partir de literais `special.type=charm` e `special.how=very`,
você pode especificar o gerador do mapa de configuração em `kustomization.yaml` como

```shell
# Criando um aruivo kustomization.yaml com o `ConfigMapGenerator`
cat <<EOF >./kustomization.yaml
configMapGenerator:
- name: special-config-2
  literals:
  - special.how=very
  - special.type=charm
EOF
```
Aplicar o diretório kustomization para criar o objeto mapa de configuração.

```shell
kubectl apply -k .
configmap/special-config-2-c92b5mmcf2 created

> ```

Reviewer: This section seems to be out-of-date with the English version.

```
SPECIAL_LEVEL
SPECIAL_TYPE
```

{{< caution >}}
Se houver alguns arquivos no diretório `/etc/config/`, eles serão excluídos.
{{< /caution >}}

{{< note >}}
Os dados de texto são expostos como arquivos, usando a codificação de caracteres UTF-8. Para usar outra codificação de caracteres, use `binaryData`.
{{< /note >}}

### Adicione dados ConfigMap para um caminho específico no volume

Use o campo `path` para especificar o caminho de arquivo desejado para ítens específicos do ConfigMap.
Nesse caso, o item `SPECIAL_LEVEL` será montado no volume `config-volume` em `/etc/config/keys`.

{{< codenew file="pods/pod-configmap-volume-specific-key.yaml" >}}

Crie o Pod:

```shell
kubectl create -f https://kubernetes.io/examples/pods/pod-configmap-volume-specific-key.yaml
```

Quando o Pod executar, o comando `cat /etc/config/keys` produz a saída abaixo:

```
very
```

{{< caution >}}
Como antes, todos os arquivos préviamente existentes no diretório `/etc/config/` serão apagados.
{{< /caution >}}

### Projete chaves para caminhos específicos e permissões de arquivos

Você pode projetar chaves para caminhos específicos e permissões específicas em uma base por-arquivo. 

> O guia do usuário [Segredos](/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod) explica a sintaxe.

Reviewer: ```suggestion
O guia do usuário [Secrets](/pt-br/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod) explica a sintaxe.
```

### Referências Opcionais

> Uma referência de ConfigMap pode ser marcada "opcional".  Se o ConfigMap for inexistente, o volume montado estará vazio. Se o ConfigMap existir, mas a chave referenciada

Reviewer: ```suggestion
Uma referência de ConfigMap pode ser marcada _opcional_.  Se o ConfigMap for inexistente, o volume montado estará vazio. Se o ConfigMap existir, mas a chave referenciada
```

é inexistente, o caminho estará ausente no ponto de montagem.

> ### mapas de configuração montados são atualizados automaticamente

Reviewer: ```suggestion
### ConfigMaps montados são atualizados automaticamente
```

> Quando um ConfigMap montado é atualizado, o conteúdo projetado é eventualmente atualizado também.  Isso se aplica no caso em que um ConfigMap referenciado opcionalmente, passe a existir após o Pod ser iniciado.

Reviewer: ```suggestion
Quando um ConfigMap montado é atualizado, o conteúdo projetado é eventualmente atualizado também.  Isso se aplica no caso em que um ConfigMap referenciado opcionalmente passe a existir após o Pod ser iniciado.
```

O Kubelet verifica se o ConfigMap montado está atualizado em cada sincronização periódica. No entanto, ele usa seu cache local baseado em TTL para obter o valor atual do ConfigMap. Como resultado, o atraso total, desde o momento em que o ConfigMap foi atualizado até o momento em que novas chaves são projetadas para o pod, pode ser tão longo quanto o
período de sincronização do Kubelet (1 minuto por padrão) + TTL de cache do ConfigMap (1 minuto por padrão) no kubelet.

{{< note >}}

> Um contêiner usando um ConfigMap como um [sub caminho](/docs/concepts/storage/volumes/#using-subpath) o volume não receberá atualizações de configuração.

Reviewer: ```suggestion
Um contêiner que esteja utilizando um ConfigMap como um [subPath](/pt-br/docs/concepts/storage/volumes/#using-subpath) de volume não receberá atualizações de ConfigMaps.
```

{{< /note >}}

<!-- discussion -->

## Compreendendo ConfigMap e Pods

> A Api de recursos do ConfigMap armazena dados de configuração como pares de chave-valor. Os dados podem ser consumidos em Pods, ou fornecer as configurações para componentes do sistema, como controladores. O ConfigMap é similar a [Segredos](/docs/concepts/configuration/secret/), mas fornece um meio de trabalhar com `strings` que não contêm informações confidenciais. Usuários e componentes do sistema podem armazenar dados de configuração em mapas de configuração.

Reviewer: ```suggestion
O recurso da API ConfigMap armazena dados de configuração como pares de chave-valor. Os dados podem ser consumidos em Pods, ou fornecidos para componentes do sistema, como controladores. O ConfigMap é similar ao [Secret](/pt-br/docs/concepts/configuration/secret/), mas fornece um meio de trabalhar com _strings_ que não contêm informações confidenciais. Usuários e componentes do sistema podem armazenar dados de configuração em ConfigMaps.
```

{{< note >}}
Os mapas de configuração devem fazer referência a arquivos de propriedades, não substituí-los. Pense no ConfigMap como representando algo semelhante ao diretório `/etc` do Linux e seus conteúdos. Por exemplo, se você criar um [Volume Kubernetes](/docs/concepts/storage/volumes/) a partir de um ConfigMap, cada item de dados no ConfigMap é representado por um arquivo individual no volume.
{{< /note >}}

> O campo `data` do ConfigMap contem os dados de configuração. Como mostrado no exemplo abaixo, isso pode ser simples -- como propriedades individuais definidas usando `--from-literal` -- ou complexo -- como arquivos de configuração ou `blobs` JSON definidos usando `--from-file`.

Reviewer: ```suggestion
O campo `data` do ConfigMap contém os dados de configuração. Como mostrado no exemplo abaixo, estes podem ser simples (como propriedades individuais definidas usando `--from-literal`) ou complexos (como arquivos de configuração ou `blobs` JSON definidos usando `--from-file`).
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  creationTimestamp: 2016-02-18T19:14:38Z
  name: example-config
  namespace: default
data:
  # exemplo de uma propriedade simples definida usando --from-literal
  example.property.1: hello
  example.property.2: world
  # exemplo de uma propriedade complexa definida usando --from-file
  example.property.file: |-
    property.1=value-1
    property.2=value-2
    property.3=value-3
```

### Restrições

> - Você deve criar um ConfigMap antes de referenciá-lo em uma especificação de Pod (a menos que você marque o ConfigMap como "optional"). Se você referenciar um ConfigMap que não existe, O pod não vai iniciar. Da mesma forma, referências a chaves que não existem no ConfigMap, impedirá o pod de iniciar.

Reviewer: ```suggestion
- Você deve criar um ConfigMap antes de referenciá-lo em uma especificação de Pod (a menos que você marque o ConfigMap como `optional`). Se você referenciar um ConfigMap que não existe, o Pod não vai iniciar. Da mesma forma, referências a chaves que não existem no ConfigMap impedirão o Pod de iniciar.
```

> - Se você usar `envFrom` para definir variáveis de ambiente do ConfigMap, chaves que são consideradas inválidas serão ignoradas. O pod poderá começar, mas os nomes inválidos serão registrados no log de eventos (`InvalidVariableNames`). A mensagem de log lista cada chave pulada. Por exemplo:

Reviewer: ```suggestion
- Se você usar `envFrom` para definir variáveis de ambiente do ConfigMap, chaves que são consideradas inválidas serão ignoradas. O Pod poderá iniciar, mas os nomes inválidos serão registrados no log de eventos (`InvalidVariableNames`). A mensagem de log lista cada chave ignorada. Por exemplo:
```
